metagraph/src/annotation/annotation_converters.cpp

```diff
@@ -983,134 +1080,339 @@ void convert_to_row_annotator(const ColumnCompressed<std::string> &source,
+            std::filesystem::create_directories(tmp_dir);
+            auto sorted_set = std::make_unique<SSD>(num_threads, num_elements, tmp_dir,
+                                                    std::numeric_limits<uint64_t>::max());
+            targets[i].push_back(std::move(sorted_set));
+        }
+    }
+
+    traverse_anno_chunked(
+            "Compute diffs", graph, graph_fname, sources,
+            [&]() {
+                for (uint32_t source_idx = 0; source_idx < sources.size(); ++source_idx) {
+                    for (uint32_t col_idx = 0; col_idx < set_rows[source_idx].size();
+                         ++col_idx) {
+                        set_rows[source_idx][col_idx].resize(0);
+                    }
                 }
-                assert(!terminal[path[i]]);
-
-                uint64_t start_idx = diffs.size();
-                std::set_symmetric_difference(rows[i].begin(), rows[i].end(),
-                                              rows[i + 1].begin(), rows[i + 1].end(),
-                                              std::back_inserter(diffs));
-
-                // if we don't gain anything by diffing, make the node terminal
-                if (diffs.size() - start_idx >= rows[i].size()) {
-                    diffs.resize(start_idx);
-                    diffs.insert(diffs.end(), rows[i].begin(), rows[i].end());
-                    set_bit(terminal.data(), path[i], true);
-                    forced_terminal_count.fetch_add(1, std::memory_order_relaxed);
+            },
+            [&](const bit_vector &source_col, uint64_t row_idx, uint64_t chunk,
+                uint64_t source_idx, uint64_t col_idx, const std::vector<uint64_t> &succ_chunk,
+                const std::vector<uint64_t> &pred_chunk,
+                const std::vector<uint64_t> &pred_chunk_idx) {
+
+                const uint64_t chunk_idx = row_idx - chunk;
+                // check successor node and add current node if it's either terminal
+                // or if its successor is 0
+                if (terminal[row_idx] || !source_col[succ_chunk[chunk_idx]])
+                    set_rows[source_idx][col_idx].push_back(row_idx);
+
+                // check non-terminal predecessor nodes and add them if they are zero
+                for (uint64_t p_idx = pred_chunk_idx[chunk_idx];
+                     p_idx < pred_chunk_idx[chunk_idx + 1]; ++p_idx) {
+                    if (!source_col[pred_chunk[p_idx]] && !terminal[pred_chunk[p_idx]])
+                        set_rows[source_idx][col_idx].push_back(pred_chunk[p_idx]);
                 }
-                pos[anno_ids[i]] = { start_idx, diffs.size() - start_idx };
-            }
+            },
+            [&](uint64_t /* start */, uint64_t /* chunk_size */) {
+                for (uint32_t source_idx = 0; source_idx < sources.size(); ++source_idx) {
+                    for (uint32_t col_idx = 0; col_idx < set_rows[source_idx].size(); ++col_idx) {
```

Comment:
use size_t to index stl containers.

Reply:
uint32_t can be 4 billion.

In fact, we use `size_t` exactly for this purpose -- to emphasize that the max value is quite small so we don't have to use uint64_t/uint32_t.

Reply:
I know how big uint32_t can be. It's quite unusual to signal that a value is small by using largest possible type...

Reply:
size_t is not the largest possible type

Reply:
From cppreference.com:
> size_t can store the maximum size of a theoretically possible object of any type (including array).

So we are using the maximum possible size of an array to indicate that the array is small.  :)

Yes, if we start splitting hairs, we can say that uint128_t is larger, but that can never be the size of an array and it's not supported on many platforms (e.g. ARM), so I didn't even consider it.

Anyway, I really don't care one way or the other, I just found the convention amusing.

Reply:
You are considering only arrays, but this doesn't have to refer to an array. It could be a more general class.

size_t can be smaller than 64 bits and it can even be smaller than 32.

It must be at least 16 bits, so theoretically when you write `size_t`, you implicitly show that the thing you index doesn't have more than 2^16 elements, or that you index something that is supposed to be indexed by size_t, like any stl container.